lib/s3.py

```diff
@@ -40,6 +50,37 @@ def is_key_present(url):
     return get_key(url.hostname) is not None
 
 
+def sign_request(url, method='GET', checksum=SHA256_NIL, headers={}, region='any', service='S3'):
+    access_key, secret_key = get_key(url.hostname)
+
+    amzdate = time.strftime('%Y%m%dT%H%M%SZ', time.gmtime())
+
+    headers = headers.copy()  # don't modify the user's copy, or the default
+    headers.update({'host': url.hostname, 'x-amz-content-sha256': checksum, 'x-amz-date': amzdate})
+    headers_str = ''.join(f'{k}:{v}\n' for k, v in sorted(headers.items()))
+    headers_list = ';'.join(sorted(headers))
+
+    credential_scope = f'{amzdate[:8]}/{region}/{service}/aws4_request'
+    signing_key = f'AWS4{secret_key}'.encode('ascii')
+    for item in credential_scope.split('/'):
+        signing_key = hmac.new(signing_key, item.encode('ascii'), hashlib.sha256).digest()
+
+    algorithm = 'AWS4-HMAC-SHA256'
+    canonical_request = f'{method}\n{url.path}\n{url.query}\n{headers_str}\n{headers_list}\n{checksum}'
+    request_hash = hashlib.sha256(canonical_request.encode('ascii')).hexdigest()
+    string_to_sign = f'{algorithm}\n{amzdate}\n{credential_scope}\n{request_hash}'
+    signature = hmac.new(signing_key, string_to_sign.encode('ascii'), hashlib.sha256).hexdigest()
+    headers['Authorization'] = f'{algorithm} Credential={access_key}/{credential_scope},' \
+        f'SignedHeaders={headers_list},Signature={signature}'
+
+    return headers
+
+
+def sign_curl(url, **kwargs):
```

Comment:
This also needs a docstring, and perhaps a better name. This returns a partial CLI argv, and `url` is not really an URL string, it's a parsed object. kwargs needs to be documented, too, and I think it'd be safer to explicitly enumerate these.

Reply:
OK.